### ISSUE TITLE
Word ask: Ability to control summarization process

### ISSUE THREAD

**Issue by vladsud** [microsoft/FluidFramework#6260]
- [x] #7603 
- [x] #7604 
- [x] #7605 
- [x] #7606

Here is my understanding / recollection of today's discussion re requirements:
1. None of regular clients should be ever summarizers (same as disabling summaries through setting _generateSummaries_ to false and OCS server rejecting posting a summary in case malicious client overwrites it).
2. OCS (server) client needs to be a summarizer. It may crash and restart, so we can't rely on it being oldest client in quorum to achieve that goal.
3. Word would love to tightly control when summaries are produced, i.e. through call IContainerRuntime.summarize() (or equivalent) and disable all our heuristics (timers, op counters, etc.).
    - We did not talk about it, but I’d assume that having these signals might be useful, so I’d position it as being able to plug in their own adapter into the heuristic flow such that they can overwrite any of the behaviors.

The last two points can be achieved by requesting summarizer and calling run on it. The api could be better and we may have some assumptions spread out in the code that will need to be fixed, but as a general flow this is likely good-enough, as we can consider this an advanced scenario.

**Comment by zagriswo** [comment]
@vladsud thanks - I think that accurately captures the ask. The only additional note that I'd add is that I don't think we'd use the existing signals/timers as mentioned in the third requirement.

**Comment by vladsud** [comment]
@arinwt - any chance we can delegate this work to someone else on the team?
That would require some more detailed design / instructions on how to achieve it.
@agarwal-navin - mostly FYI, I feel like you might add your own wish list that is around this area based on your experience of testing GC / summarization process. Is this something you might be potentially looking soon?

**Comment by agarwal-navin** [comment]
Arin is working on adding on demand summary support to the summarizer - https://github.com/microsoft/FluidFramework/issues/6382.
We have discussed the design and I believe it should cover this scenario too (if not, it will move us closer to what they want).

**Comment by arinwt** [comment]
I think these requirements can gradually be satisfied in phases.

1. First Phase - Only server-side clients will be elected summarizers.
    - **[in-progress]** Adding on-demand summarize function. The intention is to call this on the parent container that has spawned a child summarizer client. It will pass through and request a summary from the ISummarizer of the child. Current implementation will have a `force` functionality to interrupt an ongoing summary attempt if there is one, but note that there is no way to disable the normal heuristics, so they must be set to very high thresholds as a workaround.
    - **[to-do/small]** Add another optional flag/qualifier to the client details that indicates whether a client is eligible to be elected the summarizer (parent client). Right now we just check that the client is interactive and that its type is not "summarizer". We should also check this additional flag during the eligibility check. All clients should have this set to disable themselves _except_ the special server-side client.
        - Nice to have: maybe have a one-time set flag in the .metadata indicating whether eligibility is opt-in vs. opt-out.
        - Or, maybe better: make the eligibility function itself pluggable by the runtime. Note that varying functionality over time or across clients can break summarizer election being deterministic, but it should resolve itself over time.
2. Second Phase - Disable summarize heuristics or make them pluggable.
    - Add a flag to disable the regular SummarizerHeuristics around when the summarizer client should generate summaries, relying only on on-demand summarize calls.
    - Or, maybe better: Rip out SummarizerHeuristics, and have its contract be pluggable, with the current implementation as the default behavior.
3. Third Phase - Disable summarizer election heuristics / eliminate spawning summarizer client.
    - Add a way to completely disable SummaryManager code from running (better than just disableSummaries); rely on manual requests for "_summarizer"->ISummarizer.
    - Possibly allow on-demand summarize calls to spawn (and cache) a summarizer client as needed?
    - If the client itself can promise not to have local modifications/state, maybe we let it call generateSummary or on-demand summary on ISummarizer (rather than the SummaryManager passthrough to child client).

Note that after phase 1, we almost achieve all the asks above. The main drawback being that we still run a bunch of code that we don't necessarily want to. The summarizer client itself will still run its heuristics, but we are hoping that we never hit them by setting the thresholds high enough.

After phase 2, we prevent running unnecessary code (summarizer heuristics: timers, op monitoring, etc.), and perhaps even by making the heuristics pluggable, we add flexibility to our system. There are 2 remaining perf/complexity drawbacks at this point:
1. That we still run code to figure out which client should be elected to summarize when we really already know that. Personally I think this is fine, as long as we make the eligibility pluggable or easier to use in this use-case of opt-in summarizers vs. opt-out.
2. That we still require 2 containers to actually generate a summary. If the reserved server-side summarizer client will never have local modifications, then we are unnecessarily running 2 containers. However, there's nothing guaranteeing that the summarizer client won't have local changes contractually; it would just be occasional errors or confusing(maybe incorrect?) state in the summary.

**Comment by vladsud** [comment]
@arinwt, thanks for the write up.

I wonder if simpler design is possible, like this:
1. There is a way to disable summarization process for given client (1b above)
2. Allow on-demand summarization from a container, without relying on secondary container.
   - this, of course, has constraint of no local ops, but I think these constraints have to be baked into DDSs / data stores (pending state manager?) directly, as even today incorrectly written component may result in ops being generated in summarizing client, screwing the process.

One thing to add:
Having extra container would not be great on server, as it adds a lot to COGS. At the same time current POR (at least for OCS) is that they would have C# logic, so maybe it does not matter long term.

Which makes me wonder, @zagriswo - do we need tight control here from server side? I.e. if it's not needed long term, should we simply scope this issue for to exclude clients from summarizer calculation?

**Comment by vladsud** [comment]
@zagriswo, ping, can you please clarify requirements here?

**Comment by arinwt** [comment]
Recently merged PR #6841 works towards this goal.

**Comment by arinwt** [comment]
Merged PR #6840 works towards this goal as well, by making it so that disableSummaries completely stops SummaryManager code from running.

The election code still runs, but at least ContainerRuntime can work properly without a SummaryManager.

**Comment by zagriswo** [comment]
@vladsud we need tighter control than just flagging which clients can perform summarization. We'll be monitoring other metrics that will influence when we may want to summarize, and control over timing is important.

**Comment by vladsud** [comment]
One big item to consider in this design is how / when we can get to a point where server (controlling summarization process) could avoid having two container instances in order to produce summaries. Today, the whole process assumes that there is main (interactive) and summarizer container working as a pair, where some processes controlled by main container actually control summarization process.

**Comment by vladsud** [comment]
Ok, here is my take on where we are right now and what needs to be done.

First, what we have today:
1. Summarizer.run() already takes options argument that has ISummarizerOptions.**disableHeuristics** property - a way to disable heuristics runs and thus be able to manually control when summaries are happening.
   - if needed, we can refactor code and expose SummarizeHeuristicRunner in a form where external code can leverage it for signaling (i.e. take into account date points it collects and exposes), but not actually let it drive the process.
2. **Summarizer.summarizeOnDemand**() is exposed to control summaries.
3. In https://github.com/microsoft/FluidFramework/pull/7244, I made Summarizer object to be created only by summarizer client (i.e. main interactive client does not create it).
4. IContainerRuntimeOptions.summaryOptions.**generateSummaries** can control if given client 
   - options are passed to ContainerRuntime.load(), this is controlled by runtime factory. An example would be BaseContainerRuntimeFactory's constructor. I.e. host app has full control here.

So, in a world where server is responsible for summaries, we have right support for clients to to disable summaries and not participate. The missing piece is only how to make service end of story easy to use and performant. Though we might consider an optional support for server (ordering service) telling clients they should not attempt to elect summarizer. That way, policy is fully controlled by server.

Given that **memory consumption** is very important for server environments, we should figure out a design where there is only one container instance required on server. There can be couple approaches here:
1. Make main client summarize:
   - We undo # 3 above, or make Summarizer object created on demand in any container.
   - Expose Summarizer object through queries (probably helper function).
2. Make it easier for host to create only summarizer client and access Summarizer object without the need to create interactive client:
   - **formRequestSummarizerFn()** is exposed and can be used to construct right query to loader to load summarizer object (may need a bit refactoring to get rid of DeltaManager)

I'd prefer # 2 as it's more in line of how it works in "normal" case.

Note that ISummarizer interface is rather small and simple, so there does not seem to be a need to expose something like SummaryManager to host to control it in a world where host fully controls when summaries are generated (though see note above about SummarizeHeuristicRunner) .

One more thing that is likely missing - we should examine cancellations. In today's world, RunWhileConnectedCoordinator (created by summarizer) controls when summarizer should move to exit. Today, this is based on summarizer client losing connection. This is not something feasible for server environments, where one of the two should happen:
1. Loss of connectivity is not possible (as special driver instance is used that hides it and does not expose it to driver).
2. Connectivity loss is possible, but right direction is to reconnect, not close / reboot summarizer (controlling process can make their own decisions when to close summarizer client and move across machines, if needed).

Either way, it seems it's a great idea to refactor code and provide coordinator or coordinator factory to Summarizer object, and thus have more flexibility on how summarizer closes itself.
This should together with changes / fork of formRequestSummarizerFn() that sets [LoaderHeader.reconnect]: false

**Work item list:**
1. Fork / refactor formRequestSummarizerFn() to allow host (server) to create summarizer container and get to ISummarizer object.
   - reevaluate if summarizing container should reconnect on loss of connectivity (in server-controlled environments). Ideally - such things to be controlled by host (i.e. exposed functionality to be generic and support all the flows).
2. Refactor creation / control of RunWhileConnectedCoordinator such that alternative ICancellableSummarizerController implementation can be provided to Summarizer object and control actual policy.
3. (Optional) Expose ordering service policy to disable client to elect summarizer
4. UTs, test coverage to ensure everything works and we do not break it down the line.

**Overall**, I believe amount of work here is relatively small. For someone familiar with this code and all the layering, I'd estimate that it's couple days of actual refactoring work / adding helper functionality, and likely double / triple that to include UTs / sample flow (would be great to emulate that environment fully in UTs). So let's say **it's 2 weeks.**
For someone new to this code base, I'd likely **double that** estimate.

**Comment by pleath** [comment]
Support the case in which a summarizer and client are created on the fly and used to perform on-demand summarization. Related API's: Summarizer.create is a static method that creates a Summarizer and underlying client. Summarizer.summarizeOnDemand launches a summary using the summarizer produced by the create() method. Summarizer.stop() halts and disposes the summarizer and closes the summarizer client.
